include/nuttx/usb/usb.h

```diff
@@ -65,10 +65,10 @@
 #define MSBYTE(u16)                             ((u16) >> 8)     /* Get MS byte from uint16_t */
 #define LSBYTE(u16)                             ((u16) & 0xff)   /* Get LS byte from uint16_t */
 
-#define GETUINT16(p)                            (((uint16_t)p[1] << 8) | (uint16_t)p[0])
-#define GETUINT32(p)                            (((uint32_t)p[3] << 24) | \
-                                                 ((uint32_t)p[2] << 16) | \
-                                                 ((uint32_t)p[1] << 8) | (uint32_t)p[0])
+#define GETUINT16(p)                            ((uint16_t)((p[1] << 8) | p[0]))
```

Comment:
what is the warning here?

Reply:
Here:
```
Fix error: format specifies type 'unsigned short' but the argument has type 'int'
```
The error message is included in the commit message, please review the change patch by patch.

Reply:
is this a place of warning?
```
            uinfo("ECM_SET_PACKET_FILTER wValue: 0x%04hx, wIndex: 0x%04hx\n",
                  GETUINT16(ctrl->value), GETUINT16(ctrl->index));
```
If yes, then also good to use `PRIx16` instead of `hx`. All NuttX ports have `#define PRIx16 "x"`.

Reply:
I need to refresh in my memory how shift works on 16bit CPUs. For 32-bit and 64-bit there will be no issues for sure, but I'm a bit concerned if change of `GETUINT32` will produce the same output when run on 16-bit CPU. I mean if argument is extended to CPU word, then `p[3] << 24` and `(uint32_t)p[3] << 24` might result to a different values

Reply:
> is this a place of warning?
> 
> ```
>             uinfo("ECM_SET_PACKET_FILTER wValue: 0x%04hx, wIndex: 0x%04hx\n",
>                   GETUINT16(ctrl->value), GETUINT16(ctrl->index));
> ```
> 
> If yes, then also good to use `PRIx16` instead of `hx`. All NuttX ports have `#define PRIx16 "x"`.

Yes, but the root cause is that the return value of GETUINT16 should be uint16_t(unsigned short) not int, because compiler promote all integer smaller than `int` to `int`. The right fix is move the cast to the last.

> I need to refresh in my memory how shift works on 16bit CPUs. For 32-bit and 64-bit there will be no issues for sure, but I'm a bit concerned if change of `GETUINT32` will produce the same output when run on 16-bit CPU. I mean if argument is extended to CPU word, then `p[3] << 24` and `(uint32_t)p[3] << 24` might result to a different values, but I need to check if that is true

Ok, I restore GETUINT32 since it doesn't suffer the integer promotion.

Reply:
I think that compiler warning is a bit redundant and `uint16_t` parameter anyway will be promoted to `int` during the function call. Let me experiment a bit with variable parameters list a bit and I will come back with a wider answer. I need to renew things in memory by running some test code.

Reply:
Yes, variable parameters list or expression will promote uint16_t to int, but it isn't related to this problem. GETUINT16 should ensure the result type is uint16_t regardless the promotion in the following evaluation.

Reply:
Try running
```
#include <stdio.h>
#include <stdint.h>

#define MK16(p) (((uint16_t)p[1] << 8) | (uint16_t)p[0])
#define MK16_(p) ((uint16_t)((p[1] << 8) | p[0]))

int main()
{
    uint8_t a[2] = {1,2};
    uint8_t b[2] = {3,4};
    printf("Output1 %04hx %04hx\n", MK16(a), MK16(b));
    printf("Output2 %04hx %04hx\n", MK16_(a), MK16_(b));

    return 0;
}
```
Both will print
```
Output1 0201 0403
Output2 0201 0403
```
If the warning matters then I was expecting to see `0000` printed as a second value of a faulty case.
The `hx` matters only for `scanf` case and that is why `PRIx8` and `PRIx16` are defined to `"x"`.

Reply:
> Yes, variable parameters list or expression will promote uint16_t to int, but it isn't related to this problem. GETUINT16 should ensure the result type is uint16_t regardless whether the promotion happen in the following evaluation.

How do you ensure that? I can bet that the assembly code will put that value into a register. How can you ensure that register will be 16 bit long on 32 bit machine?

Reply:
I mean that `GETUINT32` will be placed into 64 bit register on 64 bit machine.

Reply:
let's just place type cast in front of return result as in
```
#  define __swap_uint32(n) \
    (uint32_t)(((((uint32_t)(n)) & 0x000000ffUL) << 24) | \
               ((((uint32_t)(n)) & 0x0000ff00UL) <<  8) | \
               ((((uint32_t)(n)) & 0x00ff0000UL) >>  8) | \
               ((((uint32_t)(n)) & 0xff000000UL) >> 24))
```
Do you agree?

Reply:
Done.